### PR TITLE
En 4808/end of epoch pending miniblocks

### PR DESCRIPTION
**Pull request by sasurobert**
Pending miniblocks handler.
Add miniblock header when block is committed.
Revert miniblock headers when block is reverted.
Call for all existing miniblock headers.
Clean existing map and storage.

### CODE REVIEW

File: process/interface.go [hunk 1]
```diff
@@ -365,6 +365,21 @@ type VirtualMachinesContainerFactory interface {
 	IsInterfaceNil() bool
 }
 
+// EndOfEpochHandler defines that actions which are needed by processor for end of epoch
```
Comment (BeniaminDrasovean) [inline process/interface.go]: Update comment.

File: endOfEpoch/metachain/pendingMiniBlocks.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	return &pendingMiniBlockHeaders{
+		marshalizer:         args.Marshalizer,
+		storage:             args.Storage,
+		mapMiniBlockHeaders: make(map[string]block.ShardMiniBlockHeader),
+	}, nil
+}
+
+//PendingMiniBlockHeaders will return a sorted list of ShardMiniBlockHeaders
+func (p *pendingMiniBlockHeaders) PendingMiniBlockHeaders() []block.ShardMiniBlockHeader {
+	shardMiniBlokcHeaders := make([]block.ShardMiniBlockHeader, 0)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/metachain/pendingMiniBlocks.go]: Done.

File: endOfEpoch/metachain/pendingMiniBlocks.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+			if mbHeader.SenderShardId == mbHeader.ReceiverShardId {
+				continue
+			}
+
+			// TODO: uncomment after merge with system vm
+			//if _, ok := mapMetaMiniBlockHdrs[string(mbHeader.Hash)]; ok {
+			//	continue
+			//}
+
+			if _, ok := p.mapMiniBlockHeaders[string(mbHeader.Hash)]; !ok {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/metachain/pendingMiniBlocks.go]: Done.

File: endOfEpoch/metachain/pendingMiniBlocks.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+			if mbHeader.SenderShardId == mbHeader.ReceiverShardId {
+				continue
+			}
+
+			// TODO: uncomment after merge with system vm
+			//if _, ok := mapMetaMiniBlockHdrs[string(mbHeader.Hash)]; ok {
+			//	continue
+			//}
+
+			if _, ok := p.mapMiniBlockHeaders[string(mbHeader.Hash)]; ok {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/metachain/pendingMiniBlocks.go]: Done.

File: endOfEpoch/metachain/trigger.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
+package metachain
+
+import (
+	"github.com/ElrondNetwork/elrond-go/core/check"
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/metachain/trigger.go]: Done.

File: endOfEpoch/metachain/trigger.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+func (t *trigger) Processed() {
+	t.isEndOfEpoch = false
+}
+
+// Epoch return the current epoch
+func (t *trigger) Epoch() uint32 {
+	return t.epoch
+}
+
+// ReceivedHeader saved the header into pool to verify if end-of-epoch conditions are fullfilled
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/metachain/trigger.go]: Done.

File: endOfEpoch/shardchain/trigger.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+// NewEndOfEpochTrigger creates a trigger to signal end of epoch
+func NewEndOfEpochTrigger(args *ArgsNewShardEndOfEpochTrigger) (*trigger, error) {
+	if args == nil {
+		return nil, endOfEpoch.ErrNilArgsNewShardEndOfEpochTrigger
+	}
+
+	return &trigger{}, nil
+}
+
+// IsEndOfEpoch returns true if conditions are fullfilled for end of epoch
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/shardchain/trigger.go]: Done.

File: node/defineOptions.go [hunk 8]
```diff
@@ -401,6 +403,17 @@ func WithBootstrapRoundIndex(bootstrapRoundIndex uint64) Option {
 	}
 }
 
+// WithEndOfEpochTrigger sets up an end of trigger option for the node
```
[the remark this answers is not in the document]
Comment (miiu96) [reply node/defineOptions.go]: Done.

File: process/interface.go (hunk 1 above)
Comment (miiu96) [reply process/interface.go]: Done.

File: endOfEpoch/metachain/pendingMiniBlocks_test.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
+package metachain
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/metachain/pendingMiniBlocks_test.go]: Done. A unit test was added.

File: endOfEpoch/metachain/trigger.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+func (t *trigger) ReceivedHeader(header data.HeaderHandler) {
+}
```
[the remark this answers is not in the document]
Comment (miiu96) [reply endOfEpoch/metachain/trigger.go]: This method shouldn't be implemented. Is declared just to implement the interface.

File: cmd/node/factory/structs.go [hunk 11]
```diff
@@ -1864,6 +1929,20 @@ func newMetaBlockProcessor(
...
 	}
 
+	miniBlockHeaderStore := data.Store.GetStorer(dataRetriever.MiniBlockHeaderUnit)
+	if check.IfNil(miniBlockHeaderStore) {
+		return nil, errors.New("could not create pending miniblocks handler because of empty miniblock header store")
+	}
+
+	argsPendingMiniBlocks := &metachainEndOfEpoch.ArgsPendingMiniBlocks{
+		Marshalizer: core.Marshalizer,
+		Storage:     nil,
```
[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/factory/structs.go]: Done. Storage must be initialize with miniBlockHeaderStore.

[the remark this answers is not in the document]
Comment (miiu96) [reply cmd/node/factory/structs.go]: Done. Storage must be initialized with miniBlockHeaderStore.

File: endOfEpoch/shardchain/trigger.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply endOfEpoch/shardchain/trigger.go]: Not done.

File: process/block/shardblock.go [hunk 12]
```diff
@@ -66,6 +66,7 @@ func NewShardProcessor(arguments ArgShardProcessor) (*shardProcessor, error) {
 		blockChainHook:                arguments.BlockChainHook,
 		txCoordinator:                 arguments.TxCoordinator,
 		rounder:                       arguments.Rounder,
+		endOfEpochTrigger:             arguments.EndOfEpochTrigger,
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: CreateNewHeader method has different pointer receiver: mp instead sp

File: endOfEpoch/errors.go [hunk 13]
```diff
@@ -23,8 +23,26 @@ var ErrNilArgsNewShardEndOfEpochTrigger = errors.New("nil arguments for shard en
...
+// ErrWrongTypeAssertion signals wrong type assertion
+var ErrWrongTypeAssertion = errors.New("wrong type assertion")
+
+// ErrNilMarshalizer signals that nil marshalizer has been provided
+var ErrNilMarshalizer = errors.New("nil marshalizer")
+
+// ErrNilStorage signals that nil storage has been provided
+var ErrNilStorage = errors.New("nil storage")
+
+// ErrNilHeaderHandler signal that a nil header handler has been provided
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply endOfEpoch/errors.go]: done

File: process/block/metablock.go [hunk 14]
```diff
@@ -466,6 +477,11 @@ func (mp *metaProcessor) RestoreBlockIntoPools(headerHandler data.HeaderHandler,
 		hdrHashes[i] = metaBlock.ShardInfo[i].HeaderHash
 	}
 
+	err := mp.pendingMiniBlocks.RevertHeader(metaBlock)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/metablock.go [hunk 15]
```diff
@@ -1435,6 +1483,17 @@ func (mp *metaProcessor) ApplyBodyToHeader(hdr data.HeaderHandler, bodyHandler d
 	metaHdr.RootHash = mp.getRootHash()
 	metaHdr.TxCount = getTxCount(shardInfo)
 
+	err = mp.pendingMiniBlocks.AddProcessedHeader(metaHdr)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/shardblock.go (hunk 12 above)
Comment (sasurobert) [reply process/block/shardblock.go]: done

File: process/block/metablock.go [hunk 16]
```diff
@@ -1093,6 +1109,38 @@ func (mp *metaProcessor) checkShardHeadersValidity() (map[uint32]data.HeaderHand
 		}
 	}
 
+	mp.hdrsForCurrBlock.mutHdrsForBlock.Lock()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/metablock.go]: done

**Pull request merged**